filter/lua.go

```diff
@@ -0,0 +1,342 @@
+}
+
+// NewLUA returns a new LUA filter.
+func NewLUA(cfg baker.FilterParams) (baker.Filter, error) {
+	dcfg := cfg.DecodedConfig.(*LUAConfig)
+
+	l := lua.NewState()
+	if err := l.DoFile(dcfg.Script); err != nil {
+		return nil, fmt.Errorf("can't compile lua script %q: %v", dcfg.Script, err)
+	}
+
+	registerLUATypes(l, cfg.ComponentParams)
+
+	luaFunc := l.GetGlobal(dcfg.FilterName)
+	if luaFunc.Type() == lua.LTNil {
+		return nil, fmt.Errorf("can't find lua filter %q in script %q", dcfg.FilterName, dcfg.Script)
+	}
+
+	// Preallocate the userdata we use to wrap the record passed to the filter.
+	// We can do this since a single instance of a baker filter is only ever
+	// processing a single record at a time, so we can reuse the lua userdata
+	// structure for it. This reduces allocations.
+	ud := l.NewUserData()
+	l.SetMetatable(ud, l.GetTypeMetatable(luaRecordTypeName))
+
+	f := &LUA{}
+
+	// Preallocate the lua next function passed to the filter.
+	luaNext := l.NewFunction(func(L *lua.LState) int {
+		f.next(fastcheckLuaRecord(L, 1).r)
+		return 0
+	})
+
+	f.l = l
+	f.ud = ud
+	f.luaNext = luaNext
+	f.luaFunc = luaFunc
+
+	runtime.SetFinalizer(f, func(f *LUA) { f.l.Close() })
+
+	return f, nil
+}
+
+// registerLUATypes registers, in the given lua state, some lua types
+// and utility functions useful to run a baker filter:
+//  - the record type
+//  - createRecord function (creates and returns a new record)
+//  - validateRecord function (takes a record and returns a boolean and a
+//    number), see baker.ComponentParams.ValidateRecord
+//  - fieldByName function (returns a field index given its name, or nil
+//    if the given field name doesn't exist)
+//  - fieldNames, an lua table where field names are indexed by their field
+//    field indexes
+func registerLUATypes(l *lua.LState, comp baker.ComponentParams) {
+	registerLUARecordType(l)
+
+	// Registers the 'createRecord' lua function.
+	l.SetGlobal("createRecord", l.NewFunction(func(L *lua.LState) int {
+		rec := comp.CreateRecord()
+		ud := recordToLua(l, rec)
+		L.Push(ud)
+		return 1
+	}))
+
+	l.SetGlobal("validateRecord", l.NewFunction(func(L *lua.LState) int {
+		luar := fastcheckLuaRecord(l, 1)
+		ok, fidx := comp.ValidateRecord(luar.r)
+		l.Push(lua.LBool(ok))
+		l.Push(lua.LNumber(fidx))
+		return 2
+	}))
+
+	l.SetGlobal("fieldByName", l.NewFunction(func(L *lua.LState) int {
+		fname := L.CheckString(1)
+		fidx, ok := comp.FieldByName(fname)
+		if !ok {
+			l.Push(lua.LNil)
+		} else {
+			l.Push(lua.LNumber(fidx))
+		}
+		return 1
+	}))
+
+	// Create the fieldNaames table.
+	fields := l.NewTable()
+	for fidx, fname := range comp.FieldNames {
+		fields.RawSet(lua.LNumber(fidx), lua.LString(fname))
+	}
+	l.SetGlobal("fieldNames", fields)
+}
+
+// TODO: at the moment LUA filter doesn't publish stats.
+// There are multiple ways to do it, either require the filter to update
+// the numbers of processed and filtered records, or deduce them automatically
```

Comment:
I think that this second hypothesis could be the best one. We could count how many times the next function is called, and update the `numProcessedLines` and `numFilteredLines` atomic accordingly.

Reply:
Since we don't have the optimizations anymore, implementing stats has become trivial now. it's implemented